### PR TITLE
[hail][streams] Table map partitions

### PR DESCRIPTION
**Pull request by patrick-schultz**
Based on #7681

I had to refactor `ParameterPack.newFields` a bit to get it to work with `ParameterPack.array`.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/EmitStream.scala [hunk 1]
```diff
@@ -848,6 +894,31 @@ object EmitStream {
...
+      case Ref(name, _) => if (refName == name) uses += mult
+      case Let(_, v, b) => traverse(v, 2); traverse(b, mult)
+      case ArrayMap(a, _, b) => traverse(a, mult); traverse(b, 2)
+      case ArrayFilter(a, _, b) => traverse(a, mult); traverse(b, 2)
+      case ArrayFlatMap(a, _, b) => traverse(a, mult); traverse(b, 2)
+      case ArrayLeftJoinDistinct(l, r, _, _, c, j) =>
+        traverse(l, mult); traverse(r, mult); traverse(c, 2); traverse(j, 2)
+      case ArrayScan(a, z, _, _, b) =>
+        traverse(a, mult); traverse(z, 2); traverse(b, 2)
+      case ArrayAggScan(a, _, q) =>
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/EmitStream.scala]: remove, no longer possible here. Should be RunAggScan instead.

**Comment by tpoterba** [comment]
will experiment with automatically generating TMP nodes and benchmark that.

**Comment by tpoterba** [comment]
This node can be used to group together multiple nodes with lowering implementations, and let us generate a TableValue that can go into the tail of relational functions that will take longer to lower. I suggested this as a last project for Milo because it would make it possible to see user impact from lowering work more quickly, in the hope that this would make us more excited and motivated.

**Comment by cseed** [comment]
> This node can be used to group together multiple nodes with lowering implementations
 > let us generate a TableValue that can go into the tail of relational functions that will take longer to lower

Sorry, I didn't sleep well last night and I must be slow today.  I don't think I understand either of these.  Can you give me examples?

In thinking about how things is intended to be used, I'm actually starting to formulate a different picture: what I think we want is

```
case class DistributedArray(
    contexts: IR, globals: IR, cname: String, gname: String, body: IR)
  extends TableIR
```

where DistributedArray has the same signature as CollectDistributedArray, but is a TableIR instead of a (value)IR and should be able to be rendered as an RVD.  In particular, this is something we can construct from a TableStage during the lowering process when we hit something that can't be lowered.  In LowreTableIR we'd have:

```
  case TableToTableApply(child, f) =>
    TableToTableApply(lower(child).toDistributedArray, f)
```

where lower(child) returns a TableStage.

**Comment by tpoterba** [comment]
> In my proposal, the body would have to have a stream type.

It does:
https://github.com/hail-is/hail/pull/8028/files#diff-beb69e35856952c39b4b4929ac5fb987R973

This design is very similar to what you're looking for. The right structure for the node wasn't super clear (we aren't super far along in lowering, so the needs aren't obvious) but I think the core piece that we need in any kind of lowering intermediate is code-generated iterators, which are implemented here.

**Comment by cseed** [comment]
> This design is very similar to what you're looking for.

It has the wrong input type (table).  That's why I'm confused about how it is going to used.  It can't be a bridge to lowered stuff because the input and output are both non-lowered.

Table collect (and friends) gets you from TableIR to (value)IR and bridges in one direction.  We need a node that goes in the other direction and this isn't it.

I'm not debating the need for the iterator stuff.

**Comment by tpoterba** [comment]
Closing for now. We'll reuse most of this code but not in its current manifestation.

**Comment by tpoterba** [comment]
Closing for now. We'll reuse most of this code but not in its current manifestation.

**Pull request closed without merging**